File: js/215_KthLargestElementinanArray.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var findKthLargest = function(nums, k) {
    return randomSelect(nums,0,nums.length-1,k);
};

var partition = function(nums,b,e){
    var x = nums[b];
    var i = b;
    for(var j = i+1; j <= e; ++j){
        if(nums[j] > x){
            var t = nums[++i];
            nums[i] = nums[j];
            nums[j] = t;
        }
    }
    var y = nums[i];
    nums[i] = nums[b];
    nums[b] = y;
    return i;
}

var randomSelect = function(nums,b,e,i){
    if(b === e) return nums[b];
    var m = partition(nums,b,e);
    var k = m - b + 1;
    if(k === i) return nums[m];
    else if( i < k) return randomSelect(nums,b,m-1,i);
    else return randomSelect(nums,m+1,e,i-k);
}

File: cpp/215_KthLargestElementinanArray.cpp
class Solution {
public:
    int findKthLargest(vector<int>& nums, int k) {
        return randomSelect(nums,0,nums.size()-1,k);
    }
private:
    int partition(vector<int>& nums, int b, int e){
        int x = nums[b];
        int i = b;
        for(int j = i+1; j <= e; ++j){
            if(nums[j] > x){
                swap(nums[++i],nums[j]);
            }
        }
        swap(nums[i],nums[b]);
        return i;
    }
    
    void swap(int &x, int &y){
        int t = x;
        x = y;
        y = t;
    }
    
    int randomSelect(vector<int>& nums, int b, int e, int i){
        if(b == e) return nums[b];
        int m = partition(nums,b,e);
        int k = m - b +1; //修正m
        if(k == i) return nums[m];
        else if( i < k) return randomSelect(nums,b,m-1,i);
        else return randomSelect(nums,m+1,e,i-k);
    }
};